hyperspy/learn/mva.py

```diff
@@ -1303,6 +1312,450 @@ def _estimate_elbow_position(self, curve_values):
         elbow_position = np.argmax(distance)
         return elbow_position
 
+    
+
+    def _cluster(self,
+                n_clusters,
+                scaled_data,
+				cluster_algorithm='kmeans',
+                cluster_kwargs={}
+                ):
+        """
+        Cluster analysis of a scaled data - internal 
+        
+        Parameters
+        ----------
+        n_clusters : int
+            Number of clusters to find.
+        scaled_data : numpy array - (number_of_samples,number_of_features)
+        cluster_algorithm: string "kmeans" or "agglomerative"
+             see scikit-lear documentation.  Default "kmeans"
+        cluster_kwargs
+            Additional parameters passed to the kmeans clustering algorithm. 
+            This may include `n_init`, the number of times the algorithm is 
+            restarted to optimize results.   
+    
+        """
+        if import_sklearn.sklearn_installed is False:
+            raise ImportError(
+                 'sklearn is not installed. Nothing done')
+        if cluster_algorithm not in cluster_algorithms:
+            raise ValueError("cluster_algorithm not supported or correctly_defined")
+		    
+        algorithm = cluster_algorithms[cluster_algorithm]
+        alg = algorithm(n_clusters=n_clusters,**cluster_kwargs)
+        alg.fit(scaled_data)
+        return alg
+    
+    def scale_data_for_clustering(self,
+                use_decomposition_results=True,
+                scale_to_unit_norm=True,
+                number_pca_components=None,
+                navigation_mask=None,
+                signal_mask=None,
+                mask_fill_value = 0.0):
+        """scale data for cluster analysis
+    
+        Results are stored in `learning_results`.
+    
+        Parameters
+        ----------
+        n_clusters : int
+            Number of clusters to find.
+        use_decomposition_results : bool or numpy array
+            If True (recommended) the signal's decomposition results are used
+            for clustering. Note - If this option is not used the raw data
+            is used. This can be memory intensive and is only recommened if
+            the Signal has a small `signal_dimension`.
+        scale_to_unit_norm: bool , default True
```

Comment:
Yes happy to change the name - 
When you say any scikit-learn preprocessing class - do you mean allow for standard and norm + a user supplied method  or  using getattr to allow any of the methods ?

Reply:
Yes, give the user the possibility to pass any scikit-learn preprocessing class by doing for example:
```python
from sklearn.preprocessing import RobustScaler

s.cluster_analysis(3, scaling=RobustScaler)
```
and checking that the class is supported:
```python
if inspect.isclass(scaling) and not issubclass(RobustScaler, sklearn.base.TransformerMixin):
    raise ValueError("The class provided to the scaling parameter must be a "
                     "scikit-learn preprocessing class.")
```
If we go for this, it may be worth considering adding an additional argument `scaling_kwargs` which would be a dictionary to pass to the scaler.

Reply:
Largely fixed, allows for any scaling and for scaling kwargs